keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,414 @@
+  * pods that are backed by volumes that aren't marked as `ReadWriteMany`. This would avoid breaking having multiple replicas of the same pod.
+
+* Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `SafeDetach` flag in CSIDriver is set to true. If so, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `volumeAttachement` to be deleted.
+
+* This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.
+
+* When the `external-attacher` detects the `volumeAttachement` object is being deleted, it needs to check if the `UNPUBLISH_FENCE` CSI controller capability is true. If so, it calls CSI driver's `ControllerUnpublishVolume` and sets `ControllerUnpublishVolume.fence` field to true. The CSI driver would need to ensure that no volume is being used through a mount point. If unable to determine the usage, the CSI driver can return an error without detaching the volume. The `external-attacher` would then retry with backoffs.
+
+* The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due to the shutdown. This should happen before the pods are being evicted. The "quarantine" taint indicates that new pods should not be scheduled on the node.
+
+### Handle the Return of Shutdown Node
+
+If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.
+
+The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.
+
+When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`. Kubelet volume manager reconciler will also try to reconstruct global mount for CSI volumes and call `NodeUnstageVolume`. If there isn't enough information to reconstruct global mount, Kubelet volume manager should cleanup the [global mount directory](/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount) to avoid problems when new pods are scheduled to this node again.
+
+The Pod GC Controller should also remove the `quarantine` taint from the nodes that are already cleaned up and are safe to run pods again. We need to make sure the taint is only removed after detach is finished.
+
+Note: We'll handle the temporarily partitioned node the same way as the shutdown node. Since the `quarantine` taint is applied on the node, we'll clean up everything on the node when its network connectivity is resumed before removing the taint and allowing pods to be scheduled on the node again.
+
+### Risks and Mitigations
+
+This KEP introduces changes, and if not tested correctly could result in data corruption for users.
```

Comment:
if i'm a user running a stateful workload, how do I opt-out of the new behavior proposed by this KEP? in other words, i'd rather have an administrator deal w/ the fallout resulting from a sick/stuck node vs. k8s possibly getting into a "node crash-loop" state.

e.g. the node enters a non-ready state for some reason unrelated to my stateful workload, which continues to run as normal. it's a critical application and i don't want k8s automatically taking any steps that could result in data loss. where's the policy control for this?

Reply:
The `SafeDetach` flag in `CSIDriver` should be set to `false` when deploying the CSI driver if the user does not want this behavior.

Reply:
> The `SafeDetach` flag in `CSIDriver` should be set to `false` when deploying the CSI driver if the user does not want this behavior.

OK, I guess I was wondering if it would (or should?) be possible to signal intent on a per-workload basis. It sounds like the answer is "no, it's configured for the cluster as a whole."

Reply:
Maybe we could consider adding a policy in the StatefulSet. 
https://github.com/kubernetes/kubernetes/blob/d849d9d057369121fc43aa3359059471e1ca9d1c/pkg/apis/apps/types.go#L50

Reply:
This KEP only handles a real node shutdown case now.